lightning/src/routing/router.rs

```diff
@@ -995,42 +995,78 @@ impl cmp::PartialOrd for RouteGraphNode {
 ///
 /// Used to construct a [`PathBuildingHop`] and to estimate [`EffectiveCapacity`].
 #[derive(Clone, Debug)]
-enum CandidateRouteHop<'a> {
+pub enum CandidateRouteHop<'a> {
 	/// A hop from the payer, where the outbound liquidity is known.
 	FirstHop {
+		/// The channel details of the first hop.
+		/// `get_outbound_payment_scid` is assumed to always return Some(scid) - this assumption is checked in
+		/// `find_route` method.
 		details: &'a ChannelDetails,
+		/// The node id of the payer.
+		/// Can be accessed via `source` method.
+		node_id: NodeId
 	},
 	/// A hop found in the [`ReadOnlyNetworkGraph`], where the channel capacity may be unknown.
 	PublicHop {
+		/// The channel info of the hop.
 		info: DirectedChannelInfo<'a>,
+		/// The short_channel_id of the channel.
 		short_channel_id: u64,
+		/// The node id of the current hop in route.
+		source_node_id: NodeId,
+		/// The node id of next hop in route.
+		target_node_id: NodeId,
```

Comment:
are you suggesting to `source_node_id: &'a NodeId, target_node_id: &'a NodeId` ?
then in `fn source` we would return `Candidate::PublicHop { src_node_id } = > *src_node_id.clone()` (because we are returning NodeId)

Reply:
How strongly are you feeling about these two ideas?
change to reference would require changes in the route function and that might leak to other structs/variables and get into a rabbit hole with the borrow checker.
regarding `DirectedChannelInfo` wouldnt that be a totally different approach from the first place? ie we are adding here some stuff we can use in `channel_penalty_msat` so this kind of change would mean even a bigger refactor? 

I do feel that this pr already give great value with making the `CandidateRouteHop` much more robust and simplifying the `channel_penalty_msat` signature & with access to more data.

Reply:
@TheBlueMatt thanks for the detailed explanation.

i think i misunderstood your previous comment, but now its more clear.
pushed some new commits and changed another two, trying to do what you are suggesting here.